File: LeetCode - May/findAnagram.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        
        int m = s.length();
        int n = p.length();
        
        ArrayList<Integer> ans = new ArrayList<Integer>();
        
        if(m < n)
            return ans;
        
        int[] phash = new int[26]; 
        int[] shash = new int[26]; 
        
        for(int i=0;i<n;i++){
            phash[p.charAt(i)-'a']++; 
            shash[s.charAt(i)-'a']++; 
        }
        
        for(int i=n;i<m;i++){
            if(Arrays.equals( phash , shash )) 
                ans.add(i-n); 
            
            shash[s.charAt(i-n)-'a']--; 
            shash[s.charAt(i)-'a']++; 
        }
        
        if(Arrays.equals( phash , shash ))
            ans.add(m-n); 
            
        return ans;
    }
}

File: LeetCode - May/findAnagram.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        
        m=len(s)
        n=len(p)
        answer=[]
        
        if( m==0 or m < n):
            return answer
        
        shash = [0 for i in range(26)]
        phash = [0 for i in range(26)]
        
        for i in range(n):
            shash[ord(s[i])-ord('a')]+=1
            phash[ord(p[i])-ord('a')]+=1
        
        for i in range(n,m):
            if( shash == phash ):
                answer.append(i-n)
            shash[ord(s[i-n])-ord('a')]-=1
            shash[ord(s[i])-ord('a')]+=1
        
        if( shash == phash ):
            answer.append(m-n)
        return answer
            